google_analytics/aggregation.py

```diff
@@ -92,58 +93,147 @@ def aggregate(df):
+    return df.groupby(["fullVisitorId", "date_temp"], as_index=False).agg(agg)
 
-    # Long to wide pivot - one row per visitor.
-    dynamic_columns = ['transactionRevenue', 'visits', 'transactions', 'pageviews']
-    wide = pd.pivot_table(grouped, index='fullVisitorId', columns='date', values=dynamic_columns)
 
-    # Collapse multi-index.
-    wide.columns = wide.columns.to_series().str.join('_')
-    wide.reset_index(inplace=True)
+def split_data(train, test, x_train_dates=('2016-08-01', '2017-11-30'), y_train_dates=('2017-12-01', '2018-01-31'),
+               x_test_dates=('2017-08-01', '2018-11-30'), selec_top_per=0.5, max_cat=10):
+    """A funtion to plit and preprocess the datasets
 
-    # Now let's also get the static columns.
-    static_columns = list(set(grouped.columns) - set(dynamic_columns) - {"date"})
-    static = grouped[static_columns]
+    """
+    merged = pd.concat([train, test], sort=False)
+    merged['transactions'].fillna(0, inplace=True)
+    merged['transactionRevenue'].fillna(0.0, inplace=True)
+    # Create some features
+    merged['date'] = merged['date'].apply(lambda x: pd.datetime.strptime(str(x), '%Y-%m-%d'))
+    merged['weekday'] = merged['date'].apply(lambda x: x.weekday())
+
+    # Aggregate the dataset
+    merged = group_data(merged)
+    # Reduce categories on static columns
+    OHE_reduce = ['operatingSystem', 'country', 'browser']  # noqa
+    merged = reduce_categories(merged, OHE_reduce, selec_top_per, max_cat)
+    # create OHE
+    merged = one_hot_encode_categoricals(merged)
+
+    # Split in train and test
+    x_train = merged[(merged["date"] >= x_train_dates[0]) & (merged["date"] <= x_train_dates[1])]
+    y_train = merged.loc[(merged["date"] >= y_train_dates[0]) & (merged["date"] <= y_train_dates[1]), ['transactionRevenue', 'fullVisitorId']]
+    y_train['target'] = y_train.groupby(['fullVisitorId'], as_index=False)['transactionRevenue'].sum()['transactionRevenue']
+    y_train['target'] = np.log(y_train['target'] + 1)
+    del y_train['transactionRevenue']
+    x_test = merged[(merged["date"] >= x_test_dates[0]) & (merged["date"] <= x_test_dates[1])]
+
+    # create dynamic features
+    x_train = aggregate(x_train)
+    x_train.columns = [x.replace('2016', '1') for x in x_train.columns]
+    x_train.columns = [x.replace('2017', '2') for x in x_train.columns]
+
+    x_test = aggregate(x_test)
+    x_test.columns = [x.replace('2017', '1') for x in x_test.columns]
+    x_test.columns = [x.replace('2018', '2') for x in x_test.columns]
+
+    # Guarantee that the names are the same in train and test
+    names_train = set(x_train.columns.values)
+    names_test = set(x_test.columns.values)
+    names = list(names_train.intersection(names_test))
+
+    # Guarantee the same users
+    names_x = x_train.columns.values
+    names_y = y_train.columns.values
+    merged = x_train.merge(y_train, on="fullVisitorId", how="inner")
```

Comment:
My idea is now completely wrong because I just found out that the rules changed yet again. We will be score on every user that appeared between May 2018 and October, which means we should also predict many zeroes. Apart from that I agree with your comment.